src/lib.rs

```diff
@@ -44,12 +44,26 @@ where
 
 impl<T> Merge for Vec<T>
 where
-    T: Hash + Eq,
+    T: Hash + Eq + Clone,
 {
-    fn merge(self, other: Self) -> Self {
-        let mut merged = self;
-        merged.extend(other);
-        merged.deduplicate()
+    fn merge(mut self, other: Self) -> Self {
+        use std::collections::HashMap;
+
+        let mut position_map: HashMap<T, usize> = self
```

Comment:
Creation position map, where key is the feature, so the complexity will be O(n).